Support instance template in google batch (#3630)
nextflow-io/nextflow: docs/google.rst

google.batch.network                           Set network name to attach the VM's network interface to. The value will be prefixed with global/networks/ unless it contains a /, in which case it is assumed to be a fully specified network resource URL. If unspecified, the global default network is used.
google.batch.serviceAccountEmail               Define the Google service account email to use for the pipeline execution. If not specified, the default Compute Engine service account for the project will be used.
google.batch.subnetwork                        Define the name of the subnetwork to attach the instance to must be specified here, when the specified network is configured for custom subnet creation. The value is prefixed with `regions/subnetworks/` unless it contains a `/`, in which case it is assumed to be a fully specified subnetwork resource URL.
============================================== =================

.. note::

>   Specifying an instance policy will overwrite the following options: ``cpuPlatform``, ``machineType``, `preemptible``, and ``spot``.

Reviewer: ```suggestion
  Specifying an instance policy will overwrite the following options: ``cpuPlatform``, ``machineType``, ``preemptible``, and ``spot``.
```

google.batch.allowedLocations                  Define the set of allowed locations for VMs to be provisioned. See `Google documentation <https://cloud.google.com/batch/docs/reference/rest/v1/projects.locations.jobs#locationpolicy>`_ for details (default: no restriction. Requires version ``22.12.0-edge`` or later).
google.batch.bootDiskSize                      Set the size of the virtual machine boot disk, e.g ``50.GB`` (default: none).
google.batch.cpuPlatform                       Set the minimum CPU Platform, e.g. ``'Intel Skylake'``. See `Specifying a minimum CPU Platform for VM instances <https://cloud.google.com/compute/docs/instances/specify-min-cpu-platform#specifications>`_ (default: none).
google.batch.instanceTemplate                  Set the Compute Engine instance template to create a job (default: none).

> google.batch.installGpuDrivers                  When ``true`` install the GPU drivers to the virtual machine (default: ``false``).

Reviewer: This option shouldn't be exposed to users, it should be set automatically by Nextflow if the `accelerator` directive was used

Author: If we only use an `accelerator ` option to add GPUs to an instance, we don't need an `installGpuDrivers` option. However, if we support `instanceTemplate` option, we need the one. Because there is no way to know a created instance template has GPUs or not. It totally depends how a user creates an instance template and we only know the name of the instance template. Hence, if we support an instance template option, an `installGpuDrivers` option is inevitable.

Reviewer: Okay, I will just add a note that this option is only needed when using an instance template.

> google.batch.instanceTemplate                  Set the Compute Engine instance template to create a job (default: none).
> google.batch.installGpuDrivers                  When ``true`` install the GPU drivers to the virtual machine (default: ``false``).

Reviewer: ```suggestion
google.batch.installGpuDrivers                 When ``true`` install the GPU drivers to the VM (default: ``false``, only needed when specifying an instance template).
google.batch.instanceTemplate                  Specify an instance template with which to create VMs (default: none).
```

Process definition
------------------

Processes can be defined as usual. By default, the ``cpus`` and ``memory`` directives are used to find the cheapest machine
type that is available at the current location and that fits the requested resources. If ``memory`` is not specified, 1GB of memory is allocated per CPU.

The ``machineType`` process directive can be specified in any of the following ways:

> * A comma-separated list of patterns. The pattern can contain a ``*`` to match any number of characters
>   and ``?`` to match any single character. Examples of valid patterns: ``c2-*``, ``m?-standard*``, ``n*``.

Reviewer: ```suggestion
* A comma-separated list of patterns. The pattern can contain a ``*`` to match any number of characters
  and ``?`` to match any single character. Examples of valid patterns: ``c2-*``, ``m?-standard*``, ``n*``. Requires  version ``23.02.0-edge`` or later.
```

* A predefined Google Compute Platform `machine type <https://cloud.google.com/compute/docs/machine-types>`_ or `custom machine type <https://cloud.google.com/compute/docs/instances/creating-instance-with-custom-machine-type>`_.

* An `instance template <https://cloud.google.com/compute/docs/instance-templates>`_, specified as ``template://<instance-template>``. Using an instance template
  will overwrite the ``accelerator`` directive, as well as the following Google Batch options: ``cpuPlatform``, ``preemptible``, and ``spot``. To use an instance template
  with GPUs, you must also set the ``google.batch.installGpuDrivers`` config option to ``true``. To use an instance template with Fusion, the instance template must include

>   a ``local-ssd`` disk named ``fusion`` with at least 375 GB. Requires version ``23.03.0.edge`` or later.

Reviewer: @pditommaso I put a warning here in the docs about which options are overwritten by the instance template. I could add warnings for these settings in the code.

Directives like `cpus` and `memory` can be used with the instance template. In this case, if the machine type is bigger than the requested resources then Google Batch will try to pack multiple tasks on the same VM.

Reviewer: I ran some tests and wasn't able to get Google Batch to pack tasks onto the same VM. Still, you can use `cpus` and `memory` with `machineType`. I might have been missing something else to make it work.

type that is available at the current location and that fits the requested resources. If ``memory`` is not specified, 1GB of memory is allocated per CPU.

The ``machineType`` process directive can be specified in any of the following ways:

* A comma-separated list of patterns. The pattern can contain a ``*`` to match any number of characters
  and ``?`` to match any single character. Examples of valid patterns: ``c2-*``, ``m?-standard*``, ``n*``.

* A predefined Google Compute Platform `machine type <https://cloud.google.com/compute/docs/machine-types>`_ or `custom machine type <https://cloud.google.com/compute/docs/instances/creating-instance-with-custom-machine-type>`_.

* An `instance template <https://cloud.google.com/compute/docs/instance-templates>`_, specified as ``template://<instance-template>``. Using an instance template
  will overwrite the ``accelerator`` directive, as well as the following Google Batch options: ``cpuPlatform``, ``preemptible``, and ``spot``. To use an instance template
  with GPUs, you must also set the ``google.batch.installGpuDrivers`` config option to ``true``. To use an instance template with Fusion, the instance template must include

>   a ``local-ssd`` disk with at least 375 GB. Requires version ``23.03.0.edge`` or later.

Reviewer: @jordeu Do you think this will work? Fusion adds a local SSD through the instance policy, which is ignored if an instance template is specified. So I figured, if the instance template includes an SSD disk, then Fusion should still work.

Reviewer 2: And the device name should be "fusion".

See here:
http://github.com/nextflow-io/nextflow/blob/a33a6f47e2160f476de9024b9415d63a57647479/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy#L253-L253

And how then is used to make Google Batch automatically mount it inside the container:
http://github.com/nextflow-io/nextflow/blob/a33a6f47e2160f476de9024b9415d63a57647479/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchFusionAdapter.groovy#L52-L52

Reviewer: Okay, then it will work as long as the device name is `fusion`. The volume mounts are not in the instance policy so they will work the same.

Reviewer: ```suggestion
  with GPUs, you must also set the ``google.batch.installGpuDrivers`` config option to ``true``. To use an instance template with Fusion, the instance template must include
  a ``local-ssd`` disk named ``fusion`` with at least 375 GB. Requires version ``23.03.0.edge`` or later.
```

The ``machineType`` process directive can be specified in any of the following ways:

* A comma-separated list of patterns. The pattern can contain a ``*`` to match any number of characters

>   and ``?`` to match any single character. Examples of valid patterns: ``c2-*``, ``m?-standard*``, ``n*``. Requires  version ``23.02.0-edge`` or later.

Reviewer 3: ```suggestion
  and ``?`` to match any single character. Examples of valid patterns: ``c2-*``, ``m?-standard*``, ``n*`` (requires  version ``23.02.0-edge`` or later).
```
